torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -870,21 +870,29 @@ void ProcessGroupNCCL::abort(c10::optional<std::string> abortReason) {
   abortCommsFromMap(inInitializationCommMap_, rank_, abortReason);
 }
 
-ProcessGroupNCCL::~ProcessGroupNCCL() {
+void ProcessGroupNCCL::shutdown() {
   terminateProcessGroup_.store(true);
 
+  // Abort all NCCL Communicators first, then join threads. Otherwise, the
+  // workCleanupLoop could get stuck when NCCL kernels run into errors.
+  std::string abortReason =
+      fmt::format("Process Group destroyed on rank {}", rank_);
+  abort(abortReason);
+
   workMetaListCV_.notify_one();
+}
+
+ProcessGroupNCCL::~ProcessGroupNCCL() {
+  shutdown();
+
 #ifdef ENABLE_NCCL_ERROR_CHECKING
-  ncclCommWatchdogThread_.join();
+  if (ncclCommWatchdogThread_.joinable()) {
+    ncclCommWatchdogThread_.join();
+  }
 #endif
 
   if (onCompletionHookThread_.joinable())
     onCompletionHookThread_.join();
```

Comment:
Did not include `.join()` as part of the `shutdown` method since `join()` could potentially block.